### PR TITLE
feat: Use Django textual metadata in GraphQL 

### PR DESCRIPTION
**Pull request by noelleleigh**
## Description

Automatically add Django model docstring and field `help_text` to the Strawberry type description and field descriptions, respectively.

These descriptions can be overridden by passing the `description` kwarg to `strawberry.django.type()` or `strawberry.django.field()`.

Added tests:

- `tests/test_types.py::test_field_metadata_preserved`
- `tests/test_types.py::test_field_metadata_overridden`

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [x] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: strawberry_django/type.py [hunk 1]
```diff
@@ -178,7 +184,10 @@ def process_type(
     if not hasattr(cls, "is_type_of"):
         cls.is_type_of = lambda obj, _info: isinstance(obj, (cls, model))
 
-    strawberry.type(cls, **kwargs)
+    # Get type description from either kwargs, or the model's docstring
+    description = kwargs.pop("description", cleandoc(model.__doc__))
```
Comment (bellini666) [inline strawberry_django/type.py:188]: Just to make sure we don't add an empty string here

```suggestion
    description = kwargs.pop("description", cleandoc(model.__doc__) if model.__doc__ else None)
```

File: README.md [hunk 2]
```diff
@@ -28,12 +28,14 @@ Full documentation is available under [docs](https://strawberry-graphql.github.i
...
 class Fruit(models.Model):
+    """A tasty treat"""
     name = models.CharField(max_length=20)
     color = models.ForeignKey('Color', blank=True, null=True,
             related_name='fruits', on_delete=models.CASCADE)
 
 class Color(models.Model):
-    name = models.CharField(max_length=20)
+    name = models.CharField(max_length=20,
+           help_text="field description")
```
Comment (bellini666) [inline README.md:38]:
```suggestion
    name = models.CharField(
        max_length=20,
        help_text="field description",
    )
```

File: docs/index.md [hunk 3]
```diff
@@ -25,12 +25,14 @@ Strawberry integration with Django.
...
 class Fruit(models.Model):
+    """A tasty treat"""
     name = models.CharField(max_length=20)
     color = models.ForeignKey('Color', blank=True, null=True,
             related_name='fruits', on_delete=models.CASCADE)
 
 class Color(models.Model):
-    name = models.CharField(max_length=20)
+    name = models.CharField(max_length=20,
+           help_text="field description")
```
Comment (bellini666) [inline docs/index.md:35]:
```suggestion
    name = models.CharField(
        max_length=20,
        help_text="field description",
    )
```

**Review by bellini666 — CHANGES_REQUESTED** [review]
Some suggestions.

Either way, what do you think of this @patrick91 ?

File: strawberry_django/type.py (hunk 1 above)
Comment (noelleleigh) [reply strawberry_django/type.py]: Here's my implementation: https://github.com/strawberry-graphql/strawberry-graphql-django/pull/160/commits/d52cfc7200aa7ed911f416299aa0b3a0f2535c89

File: README.md [hunk 4]
```diff
@@ -28,12 +28,16 @@ Full documentation is available under [docs](https://strawberry-graphql.github.i
...
+    """A tasty treat"""
     name = models.CharField(max_length=20)
     color = models.ForeignKey('Color', blank=True, null=True,
             related_name='fruits', on_delete=models.CASCADE)
 
 class Color(models.Model):
-    name = models.CharField(max_length=20)
+    name = models.CharField(
+      max_length=20,
+      help_text="field description",
```
Comment (bellini666) [inline README.md:39]: Maybe this was my suggestion, but it ended up with 2 spaces which is not usual for python =P

```suggestion
        max_length=20,
        help_text="field description",
```

File: docs/index.md [hunk 5]
```diff
@@ -25,12 +25,16 @@ Strawberry integration with Django.
...
+    """A tasty treat"""
     name = models.CharField(max_length=20)
     color = models.ForeignKey('Color', blank=True, null=True,
             related_name='fruits', on_delete=models.CASCADE)
 
 class Color(models.Model):
-    name = models.CharField(max_length=20)
+    name = models.CharField(
+      max_length=20,
+      help_text="field description",
```
Comment (bellini666) [inline docs/index.md:36]: Ditto

```suggestion
        max_length=20,
        help_text="field description",
```

**Review by bellini666 — CHANGES_REQUESTED** [review]

File: README.md (hunk 4 above)
Comment (noelleleigh) [reply README.md:39]: Ah yep, you're right, I don't know why I missed that inconsistency!

**Review by bellini666 — APPROVED** [review]
This looks fine by me! :)

I'll just ask a second opinion from @patrick91

**Review by patrick91 — COMMENTED** [review]
This looks great to me! I'd only make this configurable (which is what we will be doing in strawberry core too)

**Comment by bellini666** [comment]
@patrick91 any suggestions on where to add configurations for @noelleleigh to implement? Or maybe we should do that first in strawberry so that we can use it here?

**Comment by patrick91** [comment]
> @patrick91 any suggestions on where to add configurations for @noelleleigh to implement? Or maybe we should do that first in strawberry so that we can use it here?

in Strawberry core we have a config parameter on the schema, we could extend that maybe?

or use the common django way of defining settings for Strawberry django?

in core we use config on the schema so we can have multiple schemas with different configurations

**Comment by noelleleigh** [comment]
> in Strawberry core we have a config parameter on the schema, we could extend that maybe?
> 
> or use the common django way of defining settings for Strawberry django?
> 
> in core we use config on the schema so we can have multiple schemas with different configurations

Could you point me to the Strawberry core code for this? Also I don't have a preference for how to configure this, so give me your desired strategy and I'll implement it.

**Comment by patrick91** [comment]
> > in Strawberry core we have a config parameter on the schema, we could extend that maybe?
> > or use the common django way of defining settings for Strawberry django?
> > in core we use config on the schema so we can have multiple schemas with different configurations
> 
> Could you point me to the Strawberry core code for this? Also I don't have a preference for how to configure this, so give me your desired strategy and I'll implement it.

This is where we set the schema config: https://github.com/strawberry-graphql/strawberry/blob/9b3be99a765ae2fdb95c9802507e5600351c7006/strawberry/schema/schema.py#L68

And it is used in the schema converter, it might be more difficult to use in the context of strawberry django though 🤔 


There's an in-flight PR for this feature in core: https://github.com/strawberry-graphql/strawberry/pull/1645 I'm mentioning it because I don't know if it would conflict with what you're doing, I'd love to merge the PR once it is slimmed down a bit :)

**Comment by patrick91** [comment]
> > This is where we set the schema config: [strawberry-graphql/strawberry@`9b3be99`/strawberry/schema/schema.py#L68](https://github.com/strawberry-graphql/strawberry/blob/9b3be99a765ae2fdb95c9802507e5600351c7006/strawberry/schema/schema.py#L68)
> > And it is used in the schema converter, it might be more difficult to use in the context of strawberry django though 🤔
> 
> I've poked around a bit and it doesn't appear that any existing code in strawberry-graphql-django has either the schema or its config in scope. It's not clear to me how my code in the `process_type` function could even reference the schema config at all. Am I missing something?

No, you're exactly right, this is where the complexity is going to be, I don't have a good suggestion at the moment for handling this. Unfortunately having the configuration at the schema level made some implementations quite complex and it will make this one complex too.

We could try to figure out something for this, maybe like we did for the name conversion part: https://github.com/strawberry-graphql/strawberry/blob/main/strawberry/schema/name_converter.py

or we could decide that for django we can use a "global" configuration, like other framework do. What do you think? /cc @bellini666

**Comment by noelleleigh** [comment]
A Django setting would also be a simpler option

**Comment by noelleleigh** [comment]
> > A Django setting would also be a simpler option
> 
> yeah, let's go with that, we can update it later if needed 😊

Should it be a single setting that controls both type and field description inclusion? Or two settings, one for each?

```python
STRAWBERRY_DJANGO_INHERIT_DESCRIPTIONS = True
```

vs.

```python
STRAWBERRY_DJANGO_INHERIT_MODEL_DESCRIPTIONS = True
STRAWBERRY_DJANGO_INHERIT_FIELD_DESCRIPTIONS = True
```

**Comment by patrick91** [comment]
> > > A Django setting would also be a simpler option
> > 
> > 
> > yeah, let's go with that, we can update it later if needed 😊
> 
> Should it be a single setting that controls both type and field description inclusion? Or two settings, one for each?
> 
> ```python
> STRAWBERRY_DJANGO_INHERIT_DESCRIPTIONS = True
> ```
> 
> vs.
> 
> ```python
> STRAWBERRY_DJANGO_INHERIT_MODEL_DESCRIPTIONS = True
> STRAWBERRY_DJANGO_INHERIT_FIELD_DESCRIPTIONS = True
> ```
> 
> And what should their default value(s) be?

I'd do this personally:

```python
STRAWBERRY_DJANGO = {
    "FIELD_DESCRIPTION_FROM_HELP_TEXT": False,
    "TYPE_DESCRIPTION_FROM_MODEL_DOCSTRING": False,
}
```

and disabled by default for backwards compatibility and to also prevent potential information leaking 😊

**Comment by bellini666** [comment]
> or we could decide that for django we can use a "global" configuration, like other framework do. What do you think? /cc @bellini666

I like the idea! :)

We still need to think how to do that configuration for strawberry in the future though, but for django I think this is perfect.

**Comment by noelleleigh** [comment]
Alright, I've added code for new Django settings, added tests, and added documentation.

**Comment by noelleleigh** [comment]
How do those settings changes look to y'all?

File: strawberry_django/settings.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
+"""
+Code for interacting with Django settings.
+"""
+from typing import TypedDict
+
+from django.conf import settings
+
+
+class StrawberryDjangoSettings(TypedDict):
```
Comment (patrick91) [inline strawberry_django/settings.py:9]: Nice!

**Review by patrick91 — APPROVED** [review]
Looks pretty good to me! I'll leave it to @bellini666 to review and merge 😊

**Review by bellini666 — APPROVED** [review]
It's fine by me too! :)

**Pull request merged**